File: allocations/access-ci-service/src/test/java/org/apache/custos/access/ci/service/service/AuditServiceTest.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements. See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership. The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 *  specific language governing permissions and limitations
 *  under the License.
 */
package org.apache.custos.access.ci.service.service;

import org.apache.custos.access.ci.service.model.amie.AuditAction;
import org.apache.custos.access.ci.service.model.amie.AuditLogEntity;
import org.apache.custos.access.ci.service.model.amie.PacketEntity;
import org.apache.custos.access.ci.service.model.amie.ProcessingEventEntity;
import org.apache.custos.access.ci.service.repo.amie.AuditLogRepository;
import org.apache.custos.access.ci.service.repo.amie.PacketRepository;
import org.apache.custos.access.ci.service.repo.amie.ProcessingEventRepository;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Tag;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.ArgumentCaptor;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@ExtendWith(MockitoExtension.class)
@Tag("unit")
class AuditServiceTest {

    @Mock
    private AuditLogRepository auditLogRepository;

    @Mock
    private PacketRepository packetRepository;

    @Mock
    private ProcessingEventRepository processingEventRepository;

    private AuditService auditService;

    @BeforeEach
    void setUp() {
        auditService = new AuditService(auditLogRepository, packetRepository, processingEventRepository);
    }

    @Test
    void log_withEventId_shouldCreateAuditEntryWithAllFields() {
        String packetId = "packet-001";
        String eventId = "event-abc";
        PacketEntity packetProxy = new PacketEntity();
        ProcessingEventEntity eventProxy = new ProcessingEventEntity();

        when(packetRepository.getReferenceById(packetId)).thenReturn(packetProxy);
        when(processingEventRepository.getReferenceById(eventId)).thenReturn(eventProxy);
        when(auditLogRepository.save(any(AuditLogEntity.class))).thenAnswer(inv -> inv.getArgument(0));

        auditService.log(packetId, eventId, AuditAction.CREATE_PERSON, "Person", "person-123", "Created person from packet");

        ArgumentCaptor<AuditLogEntity> captor = ArgumentCaptor.forClass(AuditLogEntity.class);
        verify(auditLogRepository).save(captor.capture());

        AuditLogEntity saved = captor.getValue();
        assertThat(saved.getPacket()).isSameAs(packetProxy);
        assertThat(saved.getEvent()).isSameAs(eventProxy);
        assertThat(saved.getAction()).isEqualTo(AuditAction.CREATE_PERSON);
        assertThat(saved.getEntityType()).isEqualTo("Person");
        assertThat(saved.getEntityId()).isEqualTo("person-123");
        assertThat(saved.getSummary()).isEqualTo("Created person from packet");
        assertThat(saved.getCreatedAt()).isNotNull();
    }

    @Test
    void log_withNullEventId_shouldCreateAuditEntryWithNullEvent() {
        String packetId = "packet-002";
        PacketEntity packetProxy = new PacketEntity();

        when(packetRepository.getReferenceById(packetId)).thenReturn(packetProxy);
        when(auditLogRepository.save(any(AuditLogEntity.class))).thenAnswer(inv -> inv.getArgument(0));

        auditService.log(packetId, null, AuditAction.REPLY_SENT, "Packet", null, "Reply dispatched");

        ArgumentCaptor<AuditLogEntity> captor = ArgumentCaptor.forClass(AuditLogEntity.class);
        verify(auditLogRepository).save(captor.capture());

        AuditLogEntity saved = captor.getValue();
        assertThat(saved.getPacket()).isSameAs(packetProxy);
        assertThat(saved.getEvent()).isNull();
        assertThat(saved.getAction()).isEqualTo(AuditAction.REPLY_SENT);
        assertThat(saved.getEntityType()).isEqualTo("Packet");
        assertThat(saved.getEntityId()).isNull();
        assertThat(saved.getSummary()).isEqualTo("Reply dispatched");
    }

    @Test
    void log_withNullSummaryAndEntityId_shouldSaveEntityWithNulls() {
        String packetId = "packet-003";
        PacketEntity packetProxy = new PacketEntity();

        when(packetRepository.getReferenceById(packetId)).thenReturn(packetProxy);
        when(auditLogRepository.save(any(AuditLogEntity.class))).thenAnswer(inv -> inv.getArgument(0));

        auditService.log(packetId, null, AuditAction.TRANSACTION_COMPLETE, "Transaction", null, null);

        ArgumentCaptor<AuditLogEntity> captor = ArgumentCaptor.forClass(AuditLogEntity.class);
        verify(auditLogRepository).save(captor.capture());

        AuditLogEntity saved = captor.getValue();
        assertThat(saved.getAction()).isEqualTo(AuditAction.TRANSACTION_COMPLETE);
        assertThat(saved.getEntityId()).isNull();
        assertThat(saved.getSummary()).isNull();
    }

    @Test
    void log_usesGetReferenceByIdForPacketProxy() {
        String packetId = "packet-004";
        PacketEntity proxy = new PacketEntity();
        when(packetRepository.getReferenceById(packetId)).thenReturn(proxy);
        when(auditLogRepository.save(any(AuditLogEntity.class))).thenAnswer(inv -> inv.getArgument(0));

        auditService.log(packetId, null, AuditAction.CREATE_PROJECT, "Project", "proj-1", null);

        verify(packetRepository).getReferenceById(packetId);
    }

    @Test
    void log_usesGetReferenceByIdForEventProxy() {
        String packetId = "packet-005";
        String eventId = "event-xyz";
        when(packetRepository.getReferenceById(packetId)).thenReturn(new PacketEntity());
        when(processingEventRepository.getReferenceById(eventId)).thenReturn(new ProcessingEventEntity());
        when(auditLogRepository.save(any(AuditLogEntity.class))).thenAnswer(inv -> inv.getArgument(0));

        auditService.log(packetId, eventId, AuditAction.CREATE_MEMBERSHIP, "Membership", "m-1", null);

        verify(processingEventRepository).getReferenceById(eventId);
    }
}

File: allocations/access-amie/service/audit_service_test.go
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

package service

import (
	"context"
	"database/sql"
	"testing"

	"github.com/apache/airavata-custos/allocations/access-amie/model"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/mock"
	"github.com/stretchr/testify/require"
)

// ---------------------------------------------------------------------------
// Mock implementation
// ---------------------------------------------------------------------------

type mockAuditStore struct {
	mock.Mock
}

func (m *mockAuditStore) Save(ctx context.Context, tx *sql.Tx, a *model.AuditLog) error {
	args := m.Called(ctx, tx, a)
	return args.Error(0)
}

// ---------------------------------------------------------------------------
// AuditService.Log tests
// ---------------------------------------------------------------------------

func TestAuditLog_WithBothPacketAndEvent(t *testing.T) {
	ctx := context.Background()
	store := new(mockAuditStore)
	svc := NewAuditService(store)

	store.On("Save", ctx, mock.Anything, mock.MatchedBy(func(a *model.AuditLog) bool {
		return a.PacketID == "packet-1" &&
			a.EventID != nil && *a.EventID == "event-1" &&
			a.Action == model.AuditCreatePerson &&
			a.EntityType == "person" &&
			a.EntityID != nil && *a.EntityID == "p1" &&
			a.Summary != nil && *a.Summary == "created person"
	})).Return(nil)

	err := svc.Log(ctx, nil, "packet-1", "event-1", model.AuditCreatePerson, "person", "p1", "created person")

	require.NoError(t, err)
	store.AssertExpectations(t)
}

func TestAuditLog_WithNullEvent(t *testing.T) {
	ctx := context.Background()
	store := new(mockAuditStore)
	svc := NewAuditService(store)

	store.On("Save", ctx, mock.Anything, mock.MatchedBy(func(a *model.AuditLog) bool {
		return a.PacketID == "packet-2" &&
			a.EventID == nil // empty event ID yields nil pointer
	})).Return(nil)

	// Empty eventID -> EventID pointer should be nil
	err := svc.Log(ctx, nil, "packet-2", "", model.AuditReplySent, "packet", "pkt-2", "reply sent")

	require.NoError(t, err)
	store.AssertExpectations(t)
}

func TestAuditLog_WithNullEntityFields(t *testing.T) {
	ctx := context.Background()
	store := new(mockAuditStore)
	svc := NewAuditService(store)

	store.On("Save", ctx, mock.Anything, mock.MatchedBy(func(a *model.AuditLog) bool {
		// Both entityID and summary are empty strings -> should be nil pointers
		return a.PacketID == "packet-3" &&
			a.EntityID == nil &&
			a.Summary == nil
	})).Return(nil)

	err := svc.Log(ctx, nil, "packet-3", "event-3", model.AuditTransactionComplete, "packet", "", "")

	require.NoError(t, err)
	store.AssertExpectations(t)
}

func TestAuditLog_PropagatesStoreError(t *testing.T) {
	ctx := context.Background()
	store := new(mockAuditStore)
	svc := NewAuditService(store)

	store.On("Save", ctx, mock.Anything, mock.AnythingOfType("*model.AuditLog")).Return(assert.AnError)

	err := svc.Log(ctx, nil, "packet-err", "evt", model.AuditCreatePerson, "person", "p-err", "summary")

	require.Error(t, err)
	assert.Contains(t, err.Error(), "audit_service")
	store.AssertExpectations(t)
}
